pkg/controller/infrastructure/infraflow/ensure_utils.go

```diff
@@ -296,3 +298,98 @@ func isUserRouter(config *gcp.InfrastructureConfig) bool {
 func isUserVPC(config *gcp.InfrastructureConfig) bool {
 	return config.Networks.VPC != nil && len(config.Networks.VPC.Name) > 0
 }
+
+// isEquivalent return whether two slices are equivalent using reflect.DeepEqual
+//
+// Equivalent in this context means they contain the same elements, not necessarily in the same order.
+// In essence, this is set equality using DeepEqual.
+func isEquivalent[T any](s1, s2 []T) bool {
+	if len(s1) != len(s2) {
+		return false
+	}
+	for _, v := range s1 {
+		if !slices.ContainsFunc(s2, func(e T) bool { return reflect.DeepEqual(v, e) }) {
+			return false
+		}
+	}
+	return true
+}
+
+// shouldUpdate returns a boolean indicating whether there is a change between the two given rules
+// that would necessitate an update
+func shouldUpdate(oldRule, newRule *compute.Firewall) bool {
```

Comment:
similar logic we (aka "I") hide in the `updater` see https://github.com/gardener/gardener-extension-provider-gcp/blob/450a0e7a52f0f437ca33e426d909b06f03c0b72e/pkg/gcp/client/updater.go#L158. I am kinda convinced that the updater can contain this logic, although I would probably discuss if the client package is the best home for the updater